core/src/main/java/org/elasticsearch/http/HttpServer.java

```diff
@@ -117,4 +124,24 @@ void handleFavicon(HttpRequest request, HttpChannel channel) {
             channel.sendResponse(new BytesRestResponse(FORBIDDEN));
         }
     }
+
+    private static final class ResourceHandlingHttpChannel extends HttpChannel {
+        private final HttpChannel delegate;
+        private final HttpServerTransport transport;
+
+        public ResourceHandlingHttpChannel(HttpChannel delegate, HttpServerTransport transport) {
+            super(delegate.request(), delegate.detailedErrorsEnabled());
```

Comment:
I agree this is tricky. I would suggest making the RestChannel an interface it self and move the logic to the HttpChannel, which can be an abstract class. If you agree, can we make this little change in a different PR, get that in and merge it into this branch? This thing is getting bigger so any change we can do in isolation, we should..